만국박람회 [STEP 2] Mary, Kobe

devKobe24 opened the pull request:
# 만국박람회 [STEP 2]

# STEP 2 PR

안녕하세요 제임스(@inwoodev)! 😁
드디어 만국박람회의 STEP2 PR을 보내드립니다.
이번 리뷰도 잘 부탁드립니다!🙇‍♀️

## 고민했던 점
1️⃣ JSONDecoder를 따로 구조체로 만드는 부분에 대하여 고민했습니다.
조금이라도 유연성있고 확장성 있는 코드를 만들기 위하여 JSONDecoder를 따로 구조체로 만들었습니다.
그 과정에서 JSONDecoder를 구조체로 구현하는 과정이 조금 어려웠고 고민을 많이 했습니다.

2️⃣ segue를 이용한 출품작 상세화면 전환 및 데이터 전달
출품작 목록에서 테이블뷰 셀을 눌렀을 때 적용할 화면전환 기법 및 뷰 간 데이터 전달 방법에 대해 고민하였습니다.
결과적으로 메인 화면에서 출품작 목록으로 전환할 때 segue를 이용하였기 때문에 통일성을 위해 segue를 활용하였으며, 데이터 전달을 위해 화면 전환 이전에 호출되는 prepare 메서드를 활용하였습니다.
또한, 선택된 Cell에 해당하는 출품작 데이터만 전달하기 위해 TableView의 indexPathForSelectedRow 메서드를 활용하였습니다.

3️⃣ 뷰 컨트롤러의 이미지 뷰와 레이블에 직접 접근하여 데이터를 전달.
EntryDetailViewController의 IBOutlet으로 선언한 koreaEntryImageView와 entryDescLabel에 데이터를 전달하려고 할 때 직접 접근하여 데이터를 넘겨주려고 하였습니다.

아래는 예시 코드 입니다.
```swift!
let vc = EntryDetailViewController()
vc.koreaEntryImageView.image = UIImage(named: "photo")
vc.entryDescLabel.text = "설명"
```

그러자 앱이 크러시가 나는 상황이 발생하였습니다.
알고보니 koreaEntryImageView와 entryDescLabel는 런타임 시점에 생성되는데, 생성되어 있지 않은 koreaEntryImageView와 entryDescLabel에 접근하여 데이터를 억지로
집어 넣어주려하니 에러가 나며 앱이 크러시가 나는 것이였습니다.
그래서 저희는 시점을 맞춰주기 위해 다음과 같이 코드를 작성하였습니다.

```swift!
@IBOutlet weak var koreaEntryImageView: UIImageView!
	@IBOutlet weak var entryDescLabel: UILabel!
    
    var koreanEntryImage: UIImage?
    var entryDescription: String?
    var entryTitle: String?
	
    override func viewDidLoad() {
        super.viewDidLoad()
		
		guard let koreanEntryImage = koreanEntryImage,
              let entryDescription = entryDescription else {
            return
        }
		
		koreaEntryImageView.image = koreanEntryImage
		entryDescLabel.text = entryDescription
    }
```

## 조언을 얻고 싶은 부분

1️⃣ JSONDataNameSpace
```swift!
private let expoModel: ExpoModel? = {
    guard let dataAsset = NSDataAsset(name: JSONDataNameSpace.expoData) else {
        return nil
    }
        let decoder = JSONDecoder()
        let expo = try? decoder.decode(ExpoModel.self, from: dataAsset.data)
        return expo
    }()
```
위의 코드와 같이 작성했을 때 JSON data를 받아오지 못하는 상황이 발생하였습니다.
하지만 아래와 같이 코드를 작성 했을 때는 JSON data가 잘 받아와졌습니다.

```swift!
private let expoModel: ExpoModel? = {
    guard let dataAsset = NSDataAsset(name: "exposition_universelle_1900") else {
        return nil
    }
        let decoder = JSONDecoder()
        let expo = try? decoder.decode(ExpoModel.self, from: dataAsset.data)
        return expo
    }()
```

그래서 따로 JSONDecoder 구조체를 만들고 아래와 같이 코드를 작성하니 잘 받아와졌습니다. 
```swift!
private let expoModel: ExpoModel? = try? Decoder.decodeJSON(dataAssetName: JSONDataNameSpace.expoData)
```

이와 관련하여 왜 첫 번째 코드처럼 작성하면 JSONData를 받아올 수 없는지 조언을 얻고 싶습니다.

2️⃣ JSONDecoder
저희는 JSONDecoder의 확장성과 유연성을 위해 따로 구조체로 빼주었습니다.
JSONDecoder의 확장성과 유연성을 위해서 따로 구조체를 만들어 사용하는 것이 올바른 방법인지 궁금합니다.

3️⃣ Enum 폴더 관리
저희는 Enum 폴더를 각각 따로 관리해주고 있습니다.
다시 말해, ViewController에서 사용하는 Enum은 ViewController 하위단에 Enum 폴더를 생성하여 관리하고 있고, 
Model에서 사용하는 Enum은 Model 폴더 하위단에 Enum 폴더를 생성하여 관리하고 있습니다.
폴더 관리는 어떻게 해야하는지 궁금합니다.

MaryJo-github committed:
:sparkles: Feat: 만국박람회 첫 화면 구현.

Commit:
:recycle: Refactor: 테스트 케이스의 변수 타입 변경.

ExpoModelTests의 visitorExpectation변수 타입을 UInt로 변경하였습니다.

Commit:
:sparkles: Feat: UInt 확장 메서드 추가.

UInt 확장 및 changeToString 메서드를 추가했습니다.

Commit:
:sparkles: Feat: ExpoViewController 구현.

ExpoViewController 내부 구조를 구현하고 JSON 데이터를 파싱하여 데이터를 올바른 위치에 넣어주었습니다.

MaryJo-github committed:
:recycle: Refactor: descriptionLabel에 여백 추가.

descriptionLabel의 leading과 trailing에 20의 여백을 추가하였습니다.

MaryJo-github committed:
:sparkles: Feat: 한국의 출품작 목록 ViewController 생성.

MaryJo-github committed:
:sparkles: Feat: 한국의 출품작 custom cell 생성.

MaryJo-github committed:
:sparkles: Feat: 한국의 출품작 목록 화면 구현.

한국의 출품작 목록 화면 구현을 위해 에셋 추가 및 storyboard를 구현했습니다.

MaryJo-github committed:
:recycle: Refactor: ScrollView 위아래 여백 수정.

ScrollView의 top, bottom 제약을 safe area에서 superView로 변경해주었습니다.

Commit:
:sparkles: Feat: 출품작 상세 화면 구현.

Commit:
:recycle: Refactor: navigationBarHidden 및 BarButtonItem 추가

네비게이션 컨트롤러의 네비게이션 바를 hidden 시켰으며, UIBarButtonItem을 커스텀하게 생성하였습니다.

Commit:
:recycle: Refactor: Navigation 관련 코드 추가 및 데이터 전달.

UIBarButtonItem을 커스텀하게 생성, NavigationBar를 보이게 하였습니다. prepare 메서드를 사용하여 선택한 셀의 이미지, 설명, 이름 데이터를 다음 뷰컨트롤러로 전달하게 구현하였습니다.

Commit:
:recycle: Refactor: cell 악세서리 타입 추가.

Commit:
:sparkles: Feat: EntryDetailViewController을 Storyboad로 구현.

MaryJo-github committed:
:recycle: Refactor: prepareForReuse 메서드 재정의.

KoreaEntryTableViewCell 내부에 prepareForReuse 메서드를 재정의하였습니다.

Commit:
:sparkles: Feat: DecodeError 타입 생성.

Commit:
:sparkles: Feat: JSONDecode 구조체 생성.

Commit:
:sparkles: Feat: ViewControllerTitleNameSpace 타입 생성.

Commit:
:sparkles: Feat: JSONDataNameSpace 타입 생성.

Commit:
:recycle: Refactor: final 키워드 추가, 변수명 수정.

클래스에 final 키워드를 추가했으며 koreanEntryImageName를 koreanEntryImage로 변경하였습니다.

Commit:
:recycle: Refactor: final, JSONDecoder, backBarButtonItem, NameSpace 수정.

클래스의 final 키워드를 추가, JSONDecode를 추가, backBarButtonItem을 추가, 매직 리터럴 네임 스페이스를 추가하였습니다.

Commit:
:recycle: Refactor: final, JSONDecoder, BackBarButtonItem, CellIdentifier Name Space 추가 및 수정.

클래스에 final 키워드를 추가, JSONDecoder를 추가, BackBarButtonItem 추가 CellIdentifier Name Space를 추가하였습니다.

Commit:
:recycle: Refactor: final 키워드 추가.

클래스에 fianl 키워드를 추가하였습니다.

Commit:
:recycle: Refactor: 전체적인 코드 리팩토링.

MaryJo-github committed:
:recycle: Refactor: backBarButtonItem 삭제 및 ViewController title 추가

MaryJo-github committed:
:recycle: Refactor: UInt 타입 개행 컨벤션 수정.

MaryJo-github committed:
:recycle: Refactor: 파일 창작자 수정.

MaryJo-github committed:
:recycle: Refactor: 개행 컨벤션 수정.

코드의 전면적인 개행을 수정하였습니다.

inwoodev commented on Expo1900/Expo1900/Controller/ExpoViewController.swift:
```diff
@@ -0,0 +1,52 @@
+import UIKit
+
+final class ExpoViewController: UIViewController {
+	@IBOutlet weak var titleLabel: UILabel!
+	@IBOutlet weak var visitorsLabel: UILabel!
+	@IBOutlet weak var locationLabel: UILabel!
+	@IBOutlet weak var durationLabel: UILabel!
+	@IBOutlet weak var descriptionLabel: UILabel!
```
조금 더 적절한 접근제한자를 활용해 볼 수 있는지 조금 더 고민 해 볼만한 부분입니다! :)

devKobe24 replied:
- 🙋‍♀️🙋‍♂️ 답변
private 접근 제한자를 사용하여 수정해봤습니다.

inwoodev commented on Expo1900/Expo1900/Controller/ExpoViewController.swift:
```diff
@@ -0,0 +1,52 @@
+	override func viewWillAppear(_ animated: Bool) {
+		super.viewWillAppear(animated)
+        
+        self.title = ViewControllerTitleNameSpace.expo
+		navigationController?.setNavigationBarHidden(true, animated: false)
```
모든 뷰컨트롤러에서 해당 구문은 hidden값에 변화를 주기만 할 뿐 반복되네요~ 저라면 extension을 활용 해 볼것 같아요~~(이 부분은 시간이 남으시면 한 번 생각 해 보시기를 추천합니다!)

devKobe24 replied:
- 🙋‍♀️🙋‍♂️ 답변
혹시 이 부분에 대해서 조금 더 힌트를 주실 수 있을까요?😂 UINavigationController의 extension을 활용하는 방법일까요...?

inwoodev replied:
저는 UIViewController를 확장하는 방향으로 생각 해 봤어요 :)

devKobe24 replied:
@inwoodev 
```swift
extension UIViewController {
	func barHidden(_ isHidden: Bool, animated: Bool) {
		navigationController?.setNavigationBarHidden(isHidden, animated: animated)
	}
}
```

제임스 이렇게 확장을 해보았어요~ 혹시 이런 방향을 제시해주신게 맞을까요~? 🤔

inwoodev commented on Expo1900/Expo1900/Controller/KoreaEntryListViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        entryListTableView.dataSource = self
```
해당 줄은 무엇을 뜻하나요? dataSource에게 자기 자신을 왜 할당하는 걸까요?

devKobe24 replied:
- 🙋‍♀️🙋‍♂️ 답변
해당 줄은 'KoreaEntryListViewController' 클래스가 'entryListTableView'의 데이터 소스가 될 것임을 뜻합니다.
dataSource에게 자기 자신을 할당하는 이유는 'KoreaEntryListViewController'가 'UITableViewDataSource' 프로토콜을 준수함으로써 'entryListTableView'에게 필요한 메서드를 호출할 수 있도록 하기 위해서 입니다.

inwoodev commented on Expo1900/Expo1900/Controller/KoreaEntryListViewController.swift:
```diff
@@ -0,0 +1,69 @@
+		navigationController?.setNavigationBarHidden(false, animated: false)
+	}
+	
+	override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+		guard let destination = segue.destination as? EntryDetailViewController,
+			  let selectedIndex = self.entryListTableView.indexPathForSelectedRow?.row,
+              let itemsModel = itemsModel else {
+			return
+		}
```
이 부분은 저라면 [공식문서](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621490-prepare) 를 참고해서 구현할 것 같아요 :). 

> Because segues can be triggered from multiple sources, you can use the information in the segue and sender parameters to disambiguate between different logical paths in your app. For example, if the segue originated from a table view, the sender parameter would identify the table view cell that the user tapped. You could then use that information to set the data on the destination view controller. 

혹시 만약 이 부분에 대해서 더 궁금하시다면 언제든 dm남겨주세요 :)

devKobe24 replied:
- 🙋‍♀️🙋‍♂️  답변

```swift!
 override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let destination = segue.destination as? EntryDetailViewController,
              let indexPath = entryListTableView.indexPath(for: sender as! UITableViewCell),
              let itemsModel = itemsModel else {
            return
        }
        let selectedIndex = indexPath.row
        
        destination.koreanEntryImage = UIImage(named: itemsModel[selectedIndex].imageName)
        destination.entryDescription = itemsModel[selectedIndex].desc
        destination.entryTitle = itemsModel[selectedIndex].name
    }
```
공식문서를 참고하여 위와 같은 코드를 작성해보았습니다. 
sender를 UITableViewCell로 다운캐스팅하여 indexPath를 얻도록 수정하였습니다. 
하지만 결과적으로 `EntryDetailViewController` 초기화 시 데이터를 넘겨주는 방법으로 변경하면서 화면 전환 방법을 변경하였고, 위 코드는 삭제하였습니다.

inwoodev replied:
sender와 타입 캐스팅을 활용하면 어렵게 보이는 문제도 보다 쉽게 해결할 때가 종종 있더라구요 :) 👍👍👍

inwoodev commented on Expo1900/Expo1900/Controller/KoreaEntryListViewController.swift:
```diff
@@ -0,0 +1,69 @@
+			return
+		}
+		
+		destination.koreanEntryImage = UIImage(named: itemsModel[selectedIndex].imageName)
+		destination.entryDescription = itemsModel[selectedIndex].desc
+		destination.entryTitle = itemsModel[selectedIndex].name
```
외부에서 특정 객체 내부 상태변화를 주는 것은 객체지향적으로 그리 좋아보이지 않습니다! 개선 해 볼 수 있을 것 같아요!

devKobe24 replied:
- 🙋‍♀️🙋‍♂️ 답변 
  - `EntryDetailViewController` 인스턴스 초기화 시 `ItemsModel`을 주입받아 해당 객체에서 상태를 변화시켜주는 것으로 수정하였습니다!

inwoodev commented on Expo1900/Expo1900/Controller/KoreaEntryListViewController.swift:
```diff
@@ -0,0 +1,69 @@
+	}
+}
+
+extension KoreaEntryListViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        guard let numberOfEntries = itemsModel?.count else {
+            return 0
+        }
+        
+        return numberOfEntries
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+		guard let koreaEntryCell = tableView.dequeueReusableCell(withIdentifier: CellIdentifier.entryListCell,
+                                                                 for: indexPath) as? KoreaEntryTableViewCell,
+              let itemsModel = itemsModel else {
+            return UITableViewCell()
+        }
+		
+		koreaEntryCell.entryImageView.image = UIImage(named: itemsModel[indexPath.row].imageName)
+        koreaEntryCell.titleLabel.text = itemsModel[indexPath.row].name
+        koreaEntryCell.shortDescLabel.text = itemsModel[indexPath.row].shortDesc
+        
+        return koreaEntryCell
+    }
+}
```
혹시 테이블뷰의 dataSource와 delegate에 대해서 공부하신 부분 공유 받을 수 있을까요?

대리자 개념도 공부하셨다면 공유 해 주시면 감사하겠습니다~~😁

devKobe24 replied:
- 🙋‍♀️🙋‍♂️ 답변
  **UITableViewDataSource**
  객체가 데이터를 관리하고 테이블 뷰에 셀을 제공하기 위해 채택하는 메서드입니다.
  데이터를 관리하려면 UITableViewDataSource 프로토콜을 구현하는 객체인 data source 객체를 테이블에 제공합니다.
  data source 객체는 테이블의 데이터 관련 요청에 응답합니다.
  또한 테이블의 데이터를 직접 관리하거나 앱의 다른 부분과 조정하여 해당 데이터를 관리합니다.
  
  Data source의 다른 책임은 다음과 같습니다.
  - 테이블 섹션(sections) 및 행(rows)수를 보고합니다.
  - 테이블의 각 행(row)에 대한 셀을 제공합니다.
  - 섹션 머리글(headers) 및 바닥글(footers)에 대한 제목을 제공합니다.
  - 테이블의 인덱스 구성(있는 경우)
  - 기본 데이터를 변경해야 하는 사용자 또는 테이블 시작(table-initiated) 업데이트에 응답합니다.
  
  **UITableViewDelegate**
  selection 항목 관리, 섹션 머리글(header) 및 바닥글(footer) 구성, 셀 삭제 및 재정렬(reordering), 테이블 뷰에서 기타 작업 수행 방법입니다.
  
  이 프로토콜의 다음 방법을 사용하여 아래 기능을 관리합니다.
  - 사용자 지정 머리글(header) 및 바닥글(footer) view를 만들고 관리합니다.
  - 행(rows), 머리글(header) 및 바닥글(footer)의 사용자 정의(custom) 높이를 지정합니다.
  - 더 나은 스크롤링 지원을 위해 높이 추정치(estimates)를 제공합니다.
  - 행의 내용을 들여씁니다(indent)
  - 행 선택(row selection)에 응답합니다.
  - 테이블 행(rows)에서 스와이프 및 기타 작업에 응답합니다.
  - 테이블 내용 편집을 지원합니다.

inwoodev replied:
완전 꼼꼼하게 작성하신 거 같아요! 이 부분을 잘 이해하신듯 합니다🙌🙌🙌. 추가적으로 이 뷰 라이프 사이클 과정에서 위 메서드들이 호출되는 시점 또한 파악 해 두시면 나중에 디버깅하실 때 도움이 될거라고 생각해요~

inwoodev commented on Expo1900/Expo1900/Extension/UInt+.swift:
```diff
@@ -0,0 +1,18 @@
+
+import Foundation
+
+extension UInt {
+	func changeToString() -> String? {
+		let numberFormatter = NumberFormatter()
+        
+		numberFormatter.numberStyle = .decimal
+        
+		return numberFormatter.string(for: self)
+	}
+}
```
오 extenstion을 적절히 활용하셨네요~~👍👍👍

inwoodev commented on Expo1900/Expo1900/Model/Decoder.swift:
```diff
@@ -0,0 +1,27 @@
+
+import UIKit
+
+struct Decoder {
+	static func decodeJSON<`Data`: Decodable>(dataAssetName: String) throws -> `Data` {
+		let jsonDecoder: JSONDecoder = JSONDecoder()
+		
+		guard let dataAsset = NSDataAsset(name: dataAssetName) else {
+			throw DecodeError.NotFoundFile
+		}
+		
+		do {
+            jsonDecoder.keyDecodingStrategy = .convertFromSnakeCase
+			let decodeData = try jsonDecoder.decode(`Data`.self, from: dataAsset.data)
+			
+            return decodeData
+		} catch {
+			throw DecodeError.FailedDecode
+		}
+	}
+}
```
음 지금은 json이 asset으로 주어지지만 만약 json string으로 데이터가 변경되거나 네트워크를 통해 data를 받아오게 된다면
해당 타입의 스태틱메서드 네이밍이 적절할지는 의문입니다!!🤔 또한 만약 key가 snake케이스로 주어지지 않는다면??! 이러한 측면에서 봤을 때 해당 타입은 확장성 측면에서는 제한적이라고 보여집니다!

devKobe24 replied:
- 🙋‍♀️🙋‍♂️ 답변
  - 이 메서드는 Asset으로 주어진 json만 decoding하는 메서드입니다.
  따라서 json string으로 데이터가 변경되거나 네트워크를 통해 data를 받아오게 된다면 해당 타입의 스태틱 메서드 네이밍의 문제가 아닌 메서드 자체를 새로 만들어야 한다고 생각합니다.
  또한 key가 snake 케이스로 주어지지 않을 경우에는 문제가 생길 가능성이 있다고 봅니다.
  따라서 'isSnakeCase' 라는 이름의 매개변수를 생성하여 snake case인지 아닌지를 분기 처리할 수 있도록 해결하였습니다.
  말씀하신 것처럼 이 메서드는 확장성 측면에서 제한적이라고 보여지나, 저희는 현재 프로젝트에서 `ExpoViewController`와`KoreaEntryListViewController`에서 해당 작업이 중복되므로, 중복코드를 줄이기 위해 삭제하지 않기로 결정하였습니다.

inwoodev commented on Expo1900/Expo1900/Controller/EntryDetailViewController.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+	
+	override func viewWillAppear(_ animated: Bool) {
+		super.viewWillAppear(animated)
```
super를 왜 호출해야 할까요~~? :)

devKobe24 replied:
- 🙋‍♀️🙋‍♂️ 답변
    - super는 메서드를 override할 때 부모클래스의 작업을 실행하기 위해서 호출합니다.
    - [viewWillAppear](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621510-viewwillappear#discussion) 메서드는 반드시 super를 호출해주어야합니다.
    - 이 외에 super를 반드시 호출해주어야하는 메서드에는 [viewDidAppear](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621423-viewdidappear#discussion), [viewIsAppearing](https://developer.apple.com/documentation/uikit/uiviewcontroller/4195485-viewisappearing), [viewWillDisappear](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621485-viewwilldisappear), [viewDidDisappear](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621477-viewdiddisappear) 등이 있습니다.

inwoodev replied:
공식문서를 완전 꼼꼼히 읽으셨군요! 훌륭합니다~ 특히나 init같은 경우 super 호출에 더 신경써야 합니다~ 이 부분도 파악 해 두시면 분명 도움이 될거라고 생각해요!

inwoodev reviewed (commented):
@MaryJo-github  @devKobe24  스텝 2 까지 오시느라 고생 많으셨습니다!~~

코드에 고민하신 흔적이 보이기에 대충 볼 수가 없네요 😆

몇 자 남겨봅니다. 아 그리고 코멘트도 확인 해 주세요~~!

## 고민하신 부분에 대한 제 의견:

> 1️⃣ JSONDecoder를 따로 구조체로 만드는 부분에 대하여 고민했습니다.
조금이라도 유연성있고 확장성 있는 코드를 만들기 위하여 JSONDecoder를 따로 구조체로 만들었습니다.
그 과정에서 JSONDecoder를 구조체로 구현하는 과정이 조금 어려웠고 고민을 많이 했습니다.

제네릭을 사용하신 부분에서 특별히 신경을 많이 쓰신게 눈에 보입니다 :) 해당 부분 관련해서 코멘트를 남겨 봤습니다. 저 또한 나름의 근거를 토대로 답변을 드렸습니다. 저희 근거를 통해서 같이 성장 해 봐요~ 😁

---

> 2️⃣ segue를 이용한 출품작 상세화면 전환 및 데이터 전달
출품작 목록에서 테이블뷰 셀을 눌렀을 때 적용할 화면전환 기법 및 뷰 간 데이터 전달 방법에 대해 고민하였습니다.
결과적으로 메인 화면에서 출품작 목록으로 전환할 때 segue를 이용하였기 때문에 통일성을 위해 segue를 활용하였으며, 데이터 전달을 위해 화면 전환 이전에 호출되는 prepare 메서드를 활용하였습니다.
또한, 선택된 Cell에 해당하는 출품작 데이터만 전달하기 위해 TableView의 indexPathForSelectedRow 메서드를 활용하였습니다.

이 부분도 코멘트 한 번 확인 해 주시면 감사할것 같아요. 음 근데 개인적으로 통일성 때문에 segue방식으로 목록으로 화면전환 코드를 구현하셨다는 부분은 조금 근거가 부족한것 같습니다~ 가령 화면전환을 담당하는 객체를 활용하고 이러한 부분 때문에 통일성이 필요했다라면 더 납득이 될 것 같아요! 이 말씀을 드리는 이유는 스토리보드를 활용하여 뷰컨트롤러를 instantiate하는 방법 또는 코드 베이스로 뷰컨트롤러를 생성하여 프로퍼티를 주입 받는 방법 등등 다양한 선택지가 있기 때문입니다.

---

> 3️⃣ 뷰 컨트롤러의 이미지 뷰와 레이블에 직접 접근하여 데이터를 전달.
EntryDetailViewController의 IBOutlet으로 선언한 koreaEntryImageView와 entryDescLabel에 데이터를 전달하려고 할 때 직접 접근하여 데이터를 넘겨주려고 하였습니다.

> 아래는 예시 코드 입니다.

> let vc = EntryDetailViewController()
vc.koreaEntryImageView.image = UIImage(named: "photo")
vc.entryDescLabel.text = "설명"
그러자 앱이 크러시가 나는 상황이 발생하였습니다.
알고보니 koreaEntryImageView와 entryDescLabel는 런타임 시점에 생성되는데, 생성되어 있지 않은 koreaEntryImageView와 entryDescLabel에 접근하여 데이터를 억지로
집어 넣어주려하니 에러가 나며 앱이 크러시가 나는 것이였습니다.
그래서 저희는 시점을 맞춰주기 위해 다음과 같이 코드를 작성하였습니다.
``` swift
    @IBOutlet weak var koreaEntryImageView: UIImageView!
    @IBOutlet weak var entryDescLabel: UILabel!
    
    var koreanEntryImage: UIImage?
    var entryDescription: String?
    var entryTitle: String?
	
    override func viewDidLoad() {
        super.viewDidLoad()
		
		guard let koreanEntryImage = koreanEntryImage,
              let entryDescription = entryDescription else {
            return
        }
		
		koreaEntryImageView.image = koreanEntryImage
		entryDescLabel.text = entryDescription
    }
```

훌륭하게 문제를 해결 하셨네요~ 뷰컨트롤러의 라이프사이클에 대해서 조금 더 고민하시게 된 부분이라고 생각이 드네요! 다만 

``` swift
var koreanEntryImage: UIImage?
var entryDescription: String?
```
해당 프로퍼티들은 이 뷰컨트롤러에 꼭 필요함에도 옵셔널로 선언이 되어있네요. 꼭 필요해 보이는 변수들은 옵셔널로 선언하지 않고 주입받는 방식으로 수정하여 개선해 볼 수도 있을 것 같아요!


## 조금이나마 조언을 드리자면...

> 1️⃣ JSONDataNameSpace

> private let expoModel: ExpoModel? = {
    guard let dataAsset = NSDataAsset(name: JSONDataNameSpace.expoData) else {
        return nil
    }
        let decoder = JSONDecoder()
        let expo = try? decoder.decode(ExpoModel.self, from: dataAsset.data)
        return expo
    }()
> 위의 코드와 같이 작성했을 때 JSON data를 받아오지 못하는 상황이 발생하였습니다.
하지만 아래와 같이 코드를 작성 했을 때는 JSON data가 잘 받아와졌습니다.

> private let expoModel: ExpoModel? = {
    guard let dataAsset = NSDataAsset(name: "exposition_universelle_1900") else {
        return nil
    }
        let decoder = JSONDecoder()
        let expo = try? decoder.decode(ExpoModel.self, from: dataAsset.data)
        return expo
    }()
> 그래서 따로 JSONDecoder 구조체를 만들고 아래와 같이 코드를 작성하니 잘 받아와졌습니다.

> private let expoModel: ExpoModel? = try? Decoder.decodeJSON(dataAssetName: JSONDataNameSpace.expoData)
이와 관련하여 왜 첫 번째 코드처럼 작성하면 JSONData를 받아올 수 없는지 조언을 얻고 싶습니다.

저도 궁금하네요. 혹시나 싶어서 저도 한 번 프린트문을 찍어봤는데. 정상적으로 데이터 출력이 가능했습니다. 
어떤 문제였을지 궁금하네요~
<img width="970" alt="스크린샷 2023-07-04 오후 10 37 15" src="https://github.com/yagom-academy/ios-exposition-universelle/assets/69072471/5b278c07-1a6a-48b4-9799-48fa726d07de">


---


> 2️⃣ JSONDecoder
저희는 JSONDecoder의 확장성과 유연성을 위해 따로 구조체로 빼주었습니다.
JSONDecoder의 확장성과 유연성을 위해서 따로 구조체를 만들어 사용하는 것이 올바른 방법인지 궁금합니다.

이 부분은 질문에서 답을 찾을 수 있을 것 같아요!
@devKobe24  @MaryJo-github  가 보시기에 어떤 측면에서 해당 타입이 JSONDecoder를 활용함에 있어서 좀 더 확장성있게 만드는지 또는 어떠한 측면에서 유연성을 높히는데 기여하는지 설명 해 주실 수 있으실까요? 제 생각에는 두분의 설명이 질문이자 답일듯 합니다 :)

> 3️⃣ Enum 폴더 관리
저희는 Enum 폴더를 각각 따로 관리해주고 있습니다.
다시 말해, ViewController에서 사용하는 Enum은 ViewController 하위단에 Enum 폴더를 생성하여 관리하고 있고,
Model에서 사용하는 Enum은 Model 폴더 하위단에 Enum 폴더를 생성하여 관리하고 있습니다.
폴더 관리는 어떻게 해야하는지 궁금합니다.

제 개인적인 생각으로는 업무의 효율성을 증가시키는 방향이라면 어떤식이든 괜찮다고 생각합니다. 두 분이 현재 관리하는 방향도 저는 맞다고 생각해요(두 분이 느끼시기에 개발에 도움이 된다면요 😁)

Commit:
:recycle: Refactor: init 메서드 생성으로 인한 의존성 주입 구현.

Commit:
:recycle: Refactor: 화면 전환 방법 및 데이터 전송 방법 변경, tableViewDelegate 추가.

Commit:
:recycle: Refactor: configureCell 메서드 생성.

셀의 데이터 주입을 위한 메서드를 생성하였습니다.

Commit:
:recycle: Refactor: Segue 삭제 및 StoryboardIdentifier 추가.

KoreaEntryListViewController에서 EntryDetailViewController로의 Segue를 삭제하고 EntryDetailViewController의 Storyboard Identifier를 추가하였습니다.

devKobe24 commented:
추가 답변 (@inwoodev) 
#### ✚ @devKobe24 @MaryJo-github 가 보시기에 어떤 측면에서 해당 타입이 JSONDecoder를 활용함에 있어서 좀 더 확장성있게 만드는지 또는 어떠한 측면에서 유연성을 높히는데 기여하는지 설명 해 주실 수 있으실까요? 제 생각에는 두분의 설명이 질문이자 답일듯 합니다 :)
- 🙋‍♀️🙋‍♂️ 답변
    - JSONDecoder의 경우 여러곳에서 사용되기 때문에 따로 빼주는것이 중복코드를 처리하고 유지보수에는 좋으나 확장성 측면에는 제한적이라고 보여집니다. 제네릭 사용으로 유연성을 높히는데 기여했다고 생각됩니다.

inwoodev commented on Expo1900/Expo1900/Controller/KoreaEntryListViewController.swift:
```diff
@@ -67,3 +54,19 @@ extension KoreaEntryListViewController {
+		
+		guard let itemsModel = itemsModel else { return }
+		
+		guard let entryDetailViewController = storyboard?.instantiateViewController(identifier: "EntryDetailViewController",
+											  creator: { (coder) in
+			return EntryDetailViewController(itemsModel: itemsModel[indexPath.row], coder: coder)
+		}) else {
+			return
+		}
```
creator를 이용하셨군요!! 👍👍👍

inwoodev replied:
다만 identifier를 하드코딩한게 조오금 아쉽습니다~ 하드코딩하다보면 언젠가 스펠링 실수 때문에 고생할 날이 분명 올테니까요 :(

devKobe24 replied:
@inwoodev 

```swift
guard let entryDetailViewController = storyboard?.instantiateViewController(identifier: Identifier.entryDetailViewController,
									  creator: { (coder) in
	return EntryDetailViewController(itemsModel: itemsModel[indexPath.row], coder: coder)
}) else {
	return
}
```

제임스~ 다음과 같이 매직 리터럴을 네임 스페이스를 활용하여 수정해보았습니다 :)

inwoodev commented:
두 분 너무 고생하셨습니다. 해당 스텝에서 저는 TableView와 delegate에 대한 이해가 핵심이라고 생각했는데요. 이 부분을 잘 이해하신 것 같고 설명도 훌륭하게 하셨기 때문에 다음 스텝으로 넘어가도 충분하다고 생각했어요~! 일단 머지할게요! 추가적인 수정사항은 다음 스텝에 추가 해 주세요~!!
